### DISCUSSION TITLE
SAT solver: A big time gap between solution found and solver response

### DISCUSSION THREAD

**Discussion by CADBIMDeveloper** [google/or-tools#3934] [CP-SAT questions] [answered]
Hi!

I see a big time gap between the moment when an optimal solution was found and getting the response from the solver. This is a part of the log:
```
Starting CP-SAT solver v9.7.2996
Parameters: max_time_in_seconds: 600 max_memory_in_mb: 1024 log_search_progress: true max_deterministic_time: 600 num_search_workers: 0
Setting number of workers to 8
...
#23    179.75s best:4024818 next:[358310,4024817] rnd_var_lns(d=0.53 s=50 t=0.10 p=0.50 stall=1)
#24    192.16s best:4020585 next:[358310,4020584] rnd_cst_lns(d=0.81 s=51 t=0.10 p=0.67 stall=0)
#25    197.38s best:358310 next:[]         max_lp fixed_bools:0/9233
#Done  197.91s max_lp
#Done  197.99s default_lp
#Done  198.23s pseudo_costs

Task timing
...
CpSolverResponse summary:
status: OPTIMAL
objective: 358310
best_bound: 358310
integers: 33458
booleans: 9026
conflicts: 0
branches: 12523
propagations: 7212299
integer_propagations: 38427066
restarts: 3515
lp_iterations: 101
walltime: 295.131
usertime: 295.131
deterministic_time: 40.7231
gap_integral: 249.795
solution_fingerprint: 0x8c93ebd970e71456
```

So the solver found an optimal solution after 198.23s, but wall/user time is 295.131 seconds.

What does solver do in between? Is it supposed to be so or a bug there? Do you need the model?

Remark: I currently don't provide num_workers explicitely. I tried to set it to 16 and 12, but it slows down everything...

I use:
OR-tools version: 9.7.2996
Language: .Net

**Comment by lperron** [comment, marked as answer]
can you send me the model ? (model.ExportToFile("issue3934.pb.txt");

It usually comes from a worker that fails to check the time limit enough.

On windows, It can just be the deallocation of all memory used by the workers which can be very slow.

PS: Max memory is a no-op

**Reply by CADBIMDeveloper** [reply]
The model: [issue3934.pb.txt](https://github.com/google/or-tools/files/12774796/issue3934.pb.txt)
full log: [issue3934.log.txt](https://github.com/google/or-tools/files/12774795/issue3934.log.txt)

time limit should not be reached, it's 600 seconds.

What do you mean by no-op. Is it ignored?

**Reply by CADBIMDeveloper** [reply]
Hi, @lperron ,

I did a test on Linux machine on the cloud

```
#Done   86.01s max_lp initial_propagation
...
walltime: 86.727
usertime: 86.727
```

So yes, the source of the problem is a slow deallocation of workers memory on Windows.

Thank you again!

**Comment by lperron** [comment]
It only applies to pure sat solve.

**Reply by CADBIMDeveloper** [reply]
> pure sat solve

With boolean variables only, correct?

**Comment by lperron** [comment]
Booleans, bool_and and bool_or only.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by CADBIMDeveloper** [reply]
@lperron , Thanks for the clarification!

**Comment by lperron** [comment]
Can you check the model you sent me ?

It is solved trivially.

CpSolverResponse summary:

status: OPTIMAL

objective: NA

best_bound: NA

integers: 0

booleans: 0

conflicts: 0

branches: 0

propagations: 0

integer_propagations: 0

restarts: 0

lp_iterations: 0

walltime: 0.065908

usertime: 0.065908

deterministic_time: 0.00526304

gap_integral: 0

solution_fingerprint: 0x6bc877c2ddf3b3f8
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by CADBIMDeveloper** [reply]
Try this one, please: [issue3934.pb.txt](https://github.com/google/or-tools/files/12774952/issue3934.pb.txt)


Probably, I've exported the model before setting the goal :-(

**Comment by lperron** [comment]
Much better :-)

Now, it solves fast on the main branch on my laptop

Starting CP-SAT solver v9.7.9999

Parameters: log_search_progress: true num_workers: 8


Initial optimization model 'issue3934.pb (2)': (model_fingerprint:
0x5a8f682cd12b5b3e)

#Variables: 32'912 (#ints: 32'912 in objective)

  - 32'912 in [0,2552]

#kLinearN: 20 (#terms: 128'923)


Starting presolve at 0.00s

  6.28e-03s  0.00e+00d  [DetectDominanceRelations]

  3.96e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=2
#num_dual_strengthening=1

  2.90e-05s  0.00e+00d  [ExtractEncodingFromLinear] #potential_supersets=1

[Symmetry] Graph for symmetry has 87'160 nodes and 179'703 arcs.

[Symmetry] Symmetry computation done. time: 0.00851 dtime: 0.0157199

  1.50e+00s  1.00e+00d *[Probe] #probed=1'516 #new_binary_clauses=851'511

  7.19e-02s  0.00e+00d  [MaxClique] Merged 6(9562 literals) into 6(9606
literals) at_most_ones.

  6.20e-03s  0.00e+00d  [DetectDominanceRelations]

  2.00e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=1
#num_dual_strengthening=1

  6.82e-03s  0.00e+00d  [DetectDuplicateConstraints]

  3.69e-04s  0.00e+00d  [DetectDifferentVariables]

  1.28e-03s  3.76e-04d  [DetectDominatedLinearConstraints]
#relevant_constraints=19

  9.94e-03s  1.39e-04d  [ProcessSetPPC] #relevant_constraints=22
#num_inclusions=5

  1.04e-04s  0.00e+00d  [FindAlmostIdenticalLinearConstraints]

  2.52e-03s  1.60e-03d  [FindBigHorizontalLinearOverlap] #blocks=5
#saved_nz=7'409 #linears=19

  9.73e-02s  1.00e+00d *[FindBigVerticalLinearOverlap]

  1.80e-05s  0.00e+00d  [MergeClauses]

  6.20e-03s  0.00e+00d  [DetectDominanceRelations]

  2.80e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=2
#num_dual_strengthening=1

  6.57e-03s  0.00e+00d  [DetectDominanceRelations]

  1.94e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=1
#num_dual_strengthening=1

[Symmetry] Graph for symmetry has 87'180 nodes and 181'795 arcs.

[Symmetry] Symmetry computation done. time: 0.008681 dtime: 0.0157138

  1.45e+00s  1.00e+00d *[Probe] #probed=1'354 #new_binary_clauses=748'550

  9.22e-04s  0.00e+00d  [MaxClique]

  6.54e-03s  0.00e+00d  [DetectDominanceRelations]

  2.02e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=1
#num_dual_strengthening=1

  6.94e-03s  0.00e+00d  [DetectDuplicateConstraints]

  3.41e-04s  0.00e+00d  [DetectDifferentVariables]

  1.21e-03s  3.54e-04d  [DetectDominatedLinearConstraints]
#relevant_constraints=24

  9.72e-03s  1.06e-04d  [ProcessSetPPC] #relevant_constraints=27
#num_inclusions=2

  7.10e-05s  0.00e+00d  [FindAlmostIdenticalLinearConstraints]

  2.00e-03s  1.90e-03d  [FindBigHorizontalLinearOverlap] #blocks=6
#saved_nz=5'457 #linears=24

  9.75e-02s  1.00e+00d *[FindBigVerticalLinearOverlap]

  2.83e-04s  0.00e+00d  [MergeClauses]

  6.38e-03s  0.00e+00d  [DetectDominanceRelations]

  2.74e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=2
#num_dual_strengthening=1

  6.76e-03s  0.00e+00d  [DetectDominanceRelations]

  1.93e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=1
#num_dual_strengthening=1

[Symmetry] Graph for symmetry has 87'198 nodes and 176'344 arcs.

[Symmetry] Symmetry computation done. time: 0.00875 dtime: 0.0154972

  1.44e+00s  1.00e+00d *[Probe] #probed=1'348 #new_binary_clauses=738'308

  9.37e-04s  0.00e+00d  [MaxClique]

  6.18e-03s  0.00e+00d  [DetectDominanceRelations]

  1.88e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=1
#num_dual_strengthening=1

  6.88e-03s  0.00e+00d  [DetectDuplicateConstraints]

  3.44e-04s  0.00e+00d  [DetectDifferentVariables]

  1.14e-03s  3.37e-04d  [DetectDominatedLinearConstraints]
#relevant_constraints=30

  8.60e-03s  1.04e-04d  [ProcessSetPPC] #relevant_constraints=33
#num_inclusions=2

  9.50e-05s  0.00e+00d  [FindAlmostIdenticalLinearConstraints]

  2.42e-03s  2.71e-03d  [FindBigHorizontalLinearOverlap] #blocks=6
#saved_nz=5'123 #linears=30

  9.74e-02s  1.00e+00d *[FindBigVerticalLinearOverlap]

  1.90e-05s  0.00e+00d  [MergeClauses]

  6.19e-03s  0.00e+00d  [DetectDominanceRelations]

  2.53e-02s  0.00e+00d  [PresolveToFixPoint] #num_loops=2
#num_dual_strengthening=1

  7.19e-02s  0.00e+00d  [ExpandObjective] #entries=14'570'339
#tight_variables=4'404 #tight_constraints=36


Presolve summary:

  - 0 affine relations were detected.

  - rule 'TODO dual: only one blocking constraint?' was applied 912 times.

  - rule 'exactly_one: removed literals' was applied 1 time.

  - rule 'exactly_one: singleton' was applied 1 time.

  - rule 'linear matrix: common horizontal rectangle' was applied 17 times.

  - rule 'linear: extracted at most one (max).' was applied 5 times.

  - rule 'linear: fixed or dup variables' was applied 15 times.

  - rule 'linear: positive equal one' was applied 1 time.

  - rule 'linear: reduced variable domains' was applied 43 times.

  - rule 'objective: variable not used elsewhere' was applied 186 times.

  - rule 'presolve: 1169 unused variables removed.' was applied 1 time.

  - rule 'presolve: iteration' was applied 3 times.


Presolved optimization model 'issue3934.pb (2)': (model_fingerprint:
0x325f40366dc28a13)

#Variables: 31'759 (#bools: 9'026 #ints: 22'716 in objective)

  - 9'026 Booleans in [0,1]

  - 11'246 in [0,2]

  - 2'841 in [0,3]

  - 4'402 in [0,4]

  - 693 in [0,5]

  - 1'700 in [0,6]

  - 994 in [0,8]

  - 41 in [0,10]

  - 205 in [0,11]

  - 432 in [0,12]

  - 50 in [0,13]

  - 7 in [0,15]

  - 44 in [0,16]

  - 1 in [0,23]

  - 23 in [0,30]

  - 2 in [0,32]

  - 4 in [0,35]

  - 1 in [0,42]

  - 1 in [0,50]

  - 1 in [0,52]

  - 3 in [0,63]

  - 15 in [0,71]

  - 4 in [0,84]

  - 5 in [0,105]

  - 5 in [0,126]

  - 6 in [0,252]

  - 2 in [0,371]

  - 1 in [0,850]

  - 1 in [0,1276]

  - 2 in [0,2211]

  - 1 in [0,2552]

#kAtMostOne: 6 (#literals: 9'606)

#kLinearN: 36 (#terms: 107'376)


Preloading model.

#Bound   5.21s best:inf   next:[-84340,93283843] initial_domain

[Symmetry] Graph for symmetry has 86'046 nodes and 171'227 arcs.

[Symmetry] Symmetry computation done. time: 0.008493 dtime: 0.01521

#Model   5.23s var:31759/31759 constraints:42/42


Starting search at 5.23s with 8 workers.

6 full problem subsolvers: [core, default_lp, max_lp, no_lp, pseudo_costs,
reduced_costs]

1 first solution subsolver: [fj_short_default]

8 incomplete subsolvers: [feasibility_pump, graph_arc_lns, graph_cst_lns,
graph_dec_lns, graph_var_lns, rins/rens, rnd_cst_lns, rnd_var_lns]

3 helper subsolvers: [neighborhood_helper, synchronization_agent,
update_gap_integral]

#1       5.48s best:6523306 next:[-84340,6523305] fj_short_default(batch:1
#lin_moves:106 #lin_evals:949'692 #weight_updates:5)

#2       6.18s best:6514605 next:[-84340,6514604] rnd_cst_lns(d=0.50 s=10
t=0.10 p=0.00 stall=0)

#3       6.25s best:6503105 next:[-84340,6503104] graph_var_lns(d=0.50 s=11
t=0.10 p=0.00 stall=0)

#4       6.35s best:6491163 next:[-84340,6491162] graph_arc_lns(d=0.50 s=12
t=0.10 p=0.00 stall=0)

#5       6.71s best:6486960 next:[-84340,6486959] graph_cst_lns(d=0.50 s=13
t=0.10 p=0.00 stall=0)

#6       7.57s best:6224253 next:[-84340,6224252] rnd_var_lns(d=0.29 s=15
t=0.10 p=0.00 stall=1)

#7       7.99s best:6072337 next:[-84340,6072336] rnd_cst_lns(d=0.71 s=16
t=0.10 p=1.00 stall=0)

#8       8.72s best:6050142 next:[-84340,6050141] graph_var_lns(d=0.71 s=17
t=0.10 p=1.00 stall=0)

#9       9.69s best:5973540 next:[-84340,5973539] graph_cst_lns(d=0.29 s=19
t=0.10 p=0.00 stall=0)

#10     10.04s best:5929624 next:[-84340,5929623] graph_dec_lns(d=0.29 s=20
t=0.10 p=0.00 stall=1)

#11     10.52s best:5929326 next:[-84340,5929325] rnd_var_lns(d=0.46 s=21
t=0.10 p=0.50 stall=0)

#Bound  11.21s best:5929326 next:[352362,5929325] reduced_costs
initial_propagation

#12     11.99s best:386213 next:[352362,386212] rens_pump_lns(d=0.19 s=25
t=0.10 p=0.00 stall=0)

#Bound  12.08s best:386213 next:[352863,386212] default_lp

#Bound  12.45s best:386213 next:[358310,386212] max_lp initial_propagation

#13     13.96s best:379912 next:[358310,379911] graph_arc_lns(d=0.54 s=28
t=0.10 p=0.50 stall=1)

#14     14.57s best:375709 next:[358310,375708] rins_pump_lns(d=0.28 s=31
t=0.10 p=0.33 stall=0)

#15     14.70s best:371719 next:[358310,371718] rnd_cst_lns(d=0.36 s=33
t=0.10 p=0.33 stall=0)

#16     16.49s best:358310 next:[]         default_lp fixed_bools:0/9274

#Done   16.49s default_lp

#Done   16.57s max_lp


Task timing                        n [     min,      max]      avg      dev
  time         n [     min,      max]      avg      dev    dtime

  'synchronization_agent':         0 [  0.00ns,   0.00ns]   0.00ns   0.00ns
0.00ns         0 [  0.00ns,   0.00ns]   0.00ns   0.00ns   0.00ns

    'neighborhood_helper':         0 [  0.00ns,   0.00ns]   0.00ns   0.00ns
0.00ns         0 [  0.00ns,   0.00ns]   0.00ns   0.00ns   0.00ns

             'default_lp':         1 [  11.26s,   11.26s]   11.26s   0.00ns
11.26s         1 [194.67ms, 194.67ms] 194.67ms   0.00ns 194.67ms

                   'core':         1 [  11.27s,   11.27s]   11.27s   0.00ns
11.27s         1 [   3.16s,    3.16s]    3.16s   0.00ns    3.16s

                  'no_lp':         1 [  11.26s,   11.26s]   11.26s   0.00ns
11.26s         1 [   2.15s,    2.15s]    2.15s   0.00ns    2.15s

                 'max_lp':         1 [  11.33s,   11.33s]   11.33s   0.00ns
11.33s         1 [ 72.36ms,  72.36ms]  72.36ms   0.00ns  72.36ms

          'reduced_costs':         1 [  11.34s,   11.34s]   11.34s   0.00ns
11.34s         1 [227.42ms, 227.42ms] 227.42ms   0.00ns 227.42ms

           'pseudo_costs':         1 [  11.42s,   11.42s]   11.42s   0.00ns
11.42s         1 [137.52ms, 137.52ms] 137.52ms   0.00ns 137.52ms

       'feasibility_pump':         3 [ 70.75ms,    5.71s]    3.60s    2.51s
10.79s         2 [   6.51s,    7.63s]    7.07s 560.69ms   14.13s

              'rins/rens':         5 [110.74ms, 953.71ms] 433.44ms 347.30ms
  2.17s         5 [917.75us, 101.09ms]  51.92ms  42.81ms 259.60ms

            'rnd_var_lns':         4 [ 48.87ms, 540.40ms] 316.34ms 206.38ms
  1.27s         4 [ 18.84us, 100.15ms]  57.18ms  44.11ms 228.74ms

            'rnd_cst_lns':         4 [ 76.09ms, 891.70ms] 388.82ms 315.74ms
  1.56s         4 [  6.92ms, 100.62ms]  53.94ms  46.47ms 215.74ms

          'graph_var_lns':         4 [ 67.22ms, 869.84ms] 608.72ms 316.67ms
  2.43s         4 [  2.40ms, 101.00ms]  75.12ms  42.03ms 300.46ms

          'graph_arc_lns':         4 [ 96.88ms, 764.73ms] 456.78ms 268.59ms
  1.83s         4 [  3.51ms, 100.94ms]  57.35ms  44.16ms 229.42ms

          'graph_cst_lns':         4 [ 49.22ms, 362.78ms] 197.14ms 141.41ms
788.57ms         4 [ 33.85us, 100.09ms]  50.06ms  49.99ms 200.22ms

          'graph_dec_lns':         4 [ 44.54ms, 674.40ms] 373.31ms 225.34ms
  1.49s         4 [  2.51us, 100.45ms]  75.23ms  43.43ms 300.92ms

    'update_gap_integral':         0 [  0.00ns,   0.00ns]   0.00ns   0.00ns
0.00ns         0 [  0.00ns,   0.00ns]   0.00ns   0.00ns   0.00ns


Search stats         Bools  Conflicts  Branches  Restarts  BoolPropag
IntegerPropag

     'default_lp':   9'274          0    10'648     1'375   2'849'962
14'314'891

           'core':  31'742      1'878    76'496     1'375   8'517'968
21'569'224

          'no_lp':   9'558     11'326   625'590     1'376   3'349'881
21'332'060

         'max_lp':   9'037          0    10'411     1'375   2'849'962
14'280'181

  'reduced_costs':  31'313          0    32'365     1'374   2'850'284
14'328'875

   'pseudo_costs':   9'026          0    10'383     1'375   2'849'962
14'278'869


LNS stats           Improv/Calls  Closed  Difficulty  TimeLimit

      'rins/rens':           5/5     40%        0.29       0.10

    'rnd_var_lns':           2/4     50%        0.45       0.10

    'rnd_cst_lns':           4/4     50%        0.52       0.10

  'graph_var_lns':           2/4     50%        0.60       0.10

  'graph_arc_lns':           2/4     50%        0.55       0.10

  'graph_cst_lns':           2/4     50%        0.40       0.10

  'graph_dec_lns':           1/4     25%        0.18       0.10


Solutions (16)         Num     Rank

        'default_lp':    1  [16,16]

  'fj_short_default':    1    [1,1]

     'graph_arc_lns':    2   [4,13]

     'graph_cst_lns':    2    [5,9]

     'graph_dec_lns':    1  [10,10]

     'graph_var_lns':    2    [3,8]

     'rens_pump_lns':    1  [12,12]

     'rins_pump_lns':    1  [14,14]

       'rnd_cst_lns':    3   [2,15]

       'rnd_var_lns':    2   [6,11]


Objective bounds     Num

      'default_lp':    1

  'initial_domain':    1

          'max_lp':    1

   'reduced_costs':    1


Solution repositories    Added  Queried  Ignored  Synchro

  'feasible solutions':     16       62        0       16

        'lp solutions':      9        1        0        7

                'pump':      6        4


Improving bounds shared    Num

                'max_lp':  302

                 'no_lp':    1

          'pseudo_costs':  955


CpSolverResponse summary:

status: OPTIMAL

objective: 358310

best_bound: 358310

integers: 33458

booleans: 9274

conflicts: 0

branches: 10648

propagations: 2849962

integer_propagations: 14314891

restarts: 1375

lp_iterations: 191

walltime: 17.0767

usertime: 17.0767

deterministic_time: 27.9188

gap_integral: 146.521

solution_fingerprint: 0xb1bc3297f21ab67a


Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by CADBIMDeveloper** [reply]
Hmm, interesting :-)

What kind of laptop do you use?

Mine is rather old:
Win10 on `Intel Core i7-6700HQ CPU 2.60GHz (Skylake), 1 CPU, 8 logical and 4 physical cores, 16GB RAM`

And it's 10x times slower :-(

I'll try to run the app in the docker container (ubuntu or alpine) when the app is ready, but it won't be very soon. Could the source of this issue be a .Net layer?

**Comment by lperron** [comment]
mac M1 pro.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00

**Reply by CADBIMDeveloper** [reply]
Heh, I'm sticked on Windows, making add-ons for CAD applications... I'll let you know about the experiments running the app on Linux later.

Thank you so much!

**Comment by lperron** [comment]
Note that the performance was improved since 9.7. This is the main branch.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
